src/historywork/BatchWork.cpp

```diff
@@ -0,0 +1,72 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "BatchWork.h"
+#include "catchup/CatchupManager.h"
+#include "lib/util/format.h"
+#include "main/Application.h"
+#include "util/Logging.h"
+
+namespace stellar
+{
+
+BatchWork::BatchWork(Application& app, WorkParent& parent, std::string name)
+    : Work(app, parent, fmt::format("batch-work-" + name), RETRY_NEVER)
+{
+}
+
+BatchWork::~BatchWork()
+{
+    clearChildren();
+}
+
+void
+BatchWork::onReset()
+{
+    resetIter();
+    clearChildren();
+    size_t nChildren = mApp.getConfig().MAX_CONCURRENT_SUBPROCESSES;
+    while (mChildren.size() < nChildren && hasNext())
+    {
+        yieldMoreWork();
+    }
+}
+
+void
+BatchWork::notify(std::string const& child)
+{
+    auto i = mChildren.find(child);
+    if (i == mChildren.end())
+    {
+        CLOG(WARNING, "Work")
+            << "BatchWork notified by unknown child " << child;
+        return;
+    }
+
+    assert(i->second);
+    markMetrics(*(i->second));
+
+    std::vector<std::string> done;
+    for (auto const& c : mChildren)
```

Comment:
`std::remove_if` would probably be a nicer way to implement this. If you do this, you might consider refactoring lines 29-33 into a function which is used both on line 29 and as part of this.

Reply:
`std::remove_if` is not applicable for associative containers like `std::map`.

Reply:
I did not realize that `mChildren` is associative, and you are correct that `std::remove_if` does not work for associative containers. However, there is still a better solution than the two-stage delete approach used here. `std::map::erase` returns a valid iterator to the element following the last removed element. Then the following is possible:
```c++
for (auto it = mChildren.begin(); it != mChildren.end(); )
{
    if (it->second->getState() == WORK_SUCCESS)
    {
        it = mChildren.erase(it);
        // CLOG
        // yieldMoreWork if hasNext
    }
    else
    {
        ++it;
    }
}
```
There are a few instances of this pattern in existing code, compare against `SetOptionsOpFrame::doApply` (starting at line 164) and `SCP::purgeSlots` (starting at line 81) for example.